src/sqlfluff/core/parser/lexer.py

```diff
@@ -371,6 +394,27 @@ def elements_to_segments(
                     elif source_only_slice.source_idx >= source_slice.start:
                         so_slices.append(source_only_slice)
 
+                # TODO: I don't think this code will generate substitution_start
+                # and substitution_end if a templated block evaluates to the empty
+                # string. Is this actually possible, and if so, is it an issue?
+                end_substitution_slice()
```

Comment:
is there a chance where the template variable is the last thing in the file and therefore we don't get back here to close it?

Reply:
or is that covered by the `end_substitution_slice()` at line 577?

Reply:
Yes, line 577 should cover it. I haven't explicitly tested this behavior, but that's the reason I wrote it this way -- i.e. creating the `end_substitution_slice()` function and calling it both "mid-stream" as well as at the end.

Reply:
How exactly does it know when to add the end_substitution_slice in? I get that it uses the last_substitution_slice to know it hasn't closed yet, but how does it know that all the substitution contents have been added before it gets added?